Modify rule S4158: Adapt to LaYC (#2599)
SonarSource/rspec: rules/S4158/javascript/rule.adoc

== Why is this an issue?

When a collection is empty with absolute certainty, it makes no sense to access or iterate it. Doing so can lead to unexpected behavior or errors in the code. The most common cause is that population was accidentally omitted or removed.

[source,javascript,diff-id=1,diff-type=noncompliant]
----
const strings = [];

if (strings.includes("foo")) {}  // Noncompliant: strings is always empty

for (const str of strings) {}  // Noncompliant

strings.forEach(str => doSomething(str)); // Noncompliant
----

> Make sure your code provides some way to populate the collection if their elements are to be accessed.

Reviewer: ```suggestion
Make sure that the collection is populated first if their elements are to be accessed.
```

Author: not sure about this one, implementation just checks for population methods used in the variable, but they may be called on event, request, or whatever. so we allow accessing empty arrays if there is a way to populate it

Reviewer: Yeah, I dunno how to properly express the fix to suggest. Feel free to ignore what I wrote.

Author: this case is valid for example:
```js
let globalArray = [];

router.get('/', (req,res) => {
   res.send(globalArray.forEach(element => `${element} is here`))
})

router.post('/', (req, res) => {
   globalArray.push(req.body.element);
   res.send('ok!');
})
```

Reviewer: I see. Okay, let's go with your suggestion then. Looking at this example, I am convinced.

Author: ```suggestion
"Make sure there exists a path in your code that populates your collection if their elements are to be accessed."
```

> [source,javascriptdiff-id=1,diff-type=compliant]

Reviewer: ```suggestion
[source,javascript,diff-id=1,diff-type=compliant]
```
